pkg/wireguard/agent/agent.go

```diff
@@ -174,58 +182,58 @@ func (a *Agent) RestoreFinished() error {
 	return nil
 }
 
-func (a *Agent) UpdatePeer(nodeName, pubKeyHex string,
-	nodeIPv4 net.IP, podCIDRv4 *net.IPNet,
-	nodeIPv6 net.IP, podCIDRv6 *net.IPNet) error {
+func (a *Agent) UpdatePeer(nodeName, pubKeyHex string, nodeIPv4, nodeIPv6 net.IP) error {
+	// To avoid running into a deadlock, we need to lock the IPCache before
+	// calling a.Lock(), because IPCache might try to call into
+	// OnIPIdentityCacheChange concurrently
+	ipcache.IPIdentityCache.RLock()
+	defer ipcache.IPIdentityCache.RUnlock()
 
 	a.Lock()
 	defer a.Unlock()
 
-	// Handle pubKey change
-	if prevPubKeyHex, found := a.pubKeyByNodeName[nodeName]; found && prevPubKeyHex != pubKeyHex {
-		log.WithField(logfields.NodeName, nodeName).Info("Pubkey has changed")
-		// pubKeys differ, so delete old peer
-		if err := a.deletePeerByPubKey(prevPubKeyHex); err != nil {
-			return err
+	var allowedIPs []net.IPNet = nil
+	if prev := a.peerByNodeName[nodeName]; prev != nil {
+		// Handle pubKey change
+		if prev.pubKey.String() != pubKeyHex {
+			log.WithField(logfields.NodeName, nodeName).Info("Pubkey has changed")
+			// pubKeys differ, so delete old peer
+			if err := a.deletePeerByPubKey(prev.pubKey); err != nil {
+				return err
+			}
 		}
-		delete(a.pubKeyByNodeName, nodeName)
-	}
 
-	log.WithFields(logrus.Fields{
-		logfields.NodeName:  nodeName,
-		logfields.PubKey:    pubKeyHex,
-		logfields.NodeIPv4:  nodeIPv4,
-		logfields.PodCIDRv4: podCIDRv4,
-		logfields.NodeIPv6:  nodeIPv6,
-		logfields.PodCIDRv6: podCIDRv6,
-	}).Info("Adding peer")
+		// Reuse allowedIPs from existing peer config
+		allowedIPs = prev.allowedIPs
 
-	pubKey, err := wgtypes.ParseKey(pubKeyHex)
-	if err != nil {
-		return err
+		// Handle Node IP change
+		if !prev.nodeIPv4.Equal(nodeIPv4) {
+			delete(a.nodeNameByNodeIP, prev.nodeIPv4.String())
+			allowedIPs = nil // reset allowedIPs and re-initialize below
+		}
+		if !prev.nodeIPv6.Equal(nodeIPv6) {
+			delete(a.nodeNameByNodeIP, prev.nodeIPv6.String())
+			allowedIPs = nil // reset allowedIPs and re-initialize below
+		}
 	}
 
-	allowedIPs := []net.IPNet{}
-
-	if option.Config.EnableIPv4 {
-		var peerIPNet net.IPNet
-		peerIPNet.IP = nodeIPv4
-		peerIPNet.Mask = net.IPv4Mask(255, 255, 255, 255)
-		allowedIPs = append(allowedIPs, peerIPNet)
-
-		if podCIDRv4 != nil {
-			allowedIPs = append(allowedIPs, *podCIDRv4)
+	if allowedIPs == nil {
+		// (Re-)Initialize the allowedIPs list by querying the IPCache. The
+		// allowedIPs will be updated by OnIPIdentityCacheChange after this
+		// function returns.
+		if option.Config.EnableIPv4 && nodeIPv4 != nil {
+			allowedIPs = append(allowedIPs, ipcache.IPIdentityCache.LookupByHostRLocked(nodeIPv4)...)
+			allowedIPs = append(allowedIPs, *iputil.IPToPrefix(nodeIPv4))
+		}
+		if option.Config.EnableIPv6 && nodeIPv6 != nil {
+			allowedIPs = append(allowedIPs, ipcache.IPIdentityCache.LookupByHostRLocked(nodeIPv6)...)
+			allowedIPs = append(allowedIPs, *iputil.IPToPrefix(nodeIPv6))
```

Comment:
Similarly to what is done on line 456, shouldn't we catch on duplicates here? Hmm maybe we won't have duplicates in the ipcache, so it should probably be fine.

Reply:
That IPCache only has a single entry per IP is my assumption, yes.